### PR TITLE
Implement Geolocation for adding QR codes

### PR DESCRIPTION
**Pull request by nic3-14159**

### CODE REVIEW

File: app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java [hunk 1]
```diff
@@ -70,69 +98,126 @@ public void onCallBack(Integer timesScanned) {
...
+        DB.saveScannerInfoInDB(qrCode, scannerInfo, new DB.Callback() {
+            @Override
+            public void onCallBack() {
+                saveComment();
+            }
+        });
+    }
+
+    private void saveComment() {
+        if (commentEditText.getText().toString().length() != 0){
```
Comment (real-rookie) [inline app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java]: Maybe we also need a finish() and a toast in this if() block? Currently I am not sure if the user will be redirected to the homepage if the user does not write a comment.

File: app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java [hunk 2]
```diff
@@ -39,7 +42,30 @@ public class AddQRActivity extends AppCompatActivity {
...
+    private ActivityResultLauncher<String[]> locationPermissionLauncher = registerForActivityResult(
+        new ActivityResultContracts.RequestMultiplePermissions(), result -> {
+            /* result is a map with the permission name as the key and the value being a
+            boolean indicating whether the permission was granted */
+            boolean accessCoarseLocation = result.get("android.permission.ACCESS_COARSE_LOCATION");
+            boolean accessFineLocation = result.get("android.permission.ACCESS_FINE_LOCATION");
+            // Fine location access also grants coarse location access. Work with at least coarse location permissions
+            if (!accessCoarseLocation) {
+                // Location permissions denied
+                Toast.makeText(getApplicationContext(), R.string.location_perm_failure_toast, Toast.LENGTH_SHORT).show();
```
Comment (real-rookie) [inline app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java:62]: this string does not exist

File: app/src/main/java/com/cmput301w23t40/capturetheqr/PlayerLocation.java [hunk 3]
```diff
@@ -1,25 +1,43 @@
...
+    }
     /**
      * The location of the player
      */
     private QRCode.Geolocation location;
+    private FusedLocationProviderClient locationClient;
     private ArrayList<QRCode> nearbyCodes;
 
-    public PlayerLocation() {
+    public PlayerLocation(Context context) {
```
Comment (real-rookie) [inline app/src/main/java/com/cmput301w23t40/capturetheqr/PlayerLocation.java:37]: the new parameter makes line 40 in MapActivity an error

File: app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java (hunk 1 above)
Comment (real-rookie) [reply app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java]: > 

Please ignore the last comment. I have tested it, if there is no finish() in this if() block, after the user writes a comment, the user has to click on the submit button **twice** to return to the homepage

**Review by real-rookie — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java (hunk 2 above)
Comment (nic3-14159) [reply app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java:62]: Fixed

File: app/src/main/java/com/cmput301w23t40/capturetheqr/PlayerLocation.java (hunk 3 above)
Comment (nic3-14159) [reply app/src/main/java/com/cmput301w23t40/capturetheqr/PlayerLocation.java:37]: Fixed

File: app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java (hunk 1 above)
Comment (nic3-14159) [reply app/src/main/java/com/cmput301w23t40/capturetheqr/AddQRActivity.java]: Fixed. Yeah, I was confused with your last comment as the else block would run if there was no comment, and that block did have the finish and toast. I've moved the finish and toast to a function and called it all exit paths.

**Review by real-rookie — APPROVED** [review]

**Pull request merged**